### PR TITLE
Support managing plugin keys 

### PR DESCRIPTION
**Pull request by qmuntal**
This PR adds support for:
- Listing available plugins using `notation plugin list`:
```cmd
> notation plugin list  
NAME   DESCRIPTION   VERSION   URL           SUPPORTED CONTRACTS   CAPABILITIES   ERROR                                                                      PATH
foo    friendly      1         example.com   [1]                   [cap]          executable name must be "notation-foo.exe" instead of "notation-baz.exe"   C:\Users\qmuntaldiaz\.notation\plugins\baz\notation-baz.exe
foo    friendly      1         example.com   [1]                   [cap]          <nil>                                                                      C:\Users\qmuntaldiaz\.notation\plugins\foo\notation-foo.exe
```
- Adding plugin keys using `notation key add --plugin=foo --name=foo-key --id=foo-id`.
- Listing plugin keys using `notation key list`:
```cmd
> notation key list
NAME      KEY PATH   CERTIFICATE PATH   ID       PLUGIN NAME   
foo-key                                 foo-id   foo
```
- Removing plugin keys using `notation key remove`.

Other changes:
- Simplified the types in `pkg/config` and reduced the API surface.
- Using go 1.18 Generics to operate on slices.
- Improved how we are formatting lists to the standard output.

@gokarnm @SteveLasker 

Signed-off-by: qmuntal <qmuntaldiaz@microsoft.com>

### CODE REVIEW

File: cmd/notation/key.go [hunk 1]
```diff
@@ -31,12 +36,21 @@ var (
...
 		Usage:     "Add key to signing key list",
-		ArgsUsage: "<key_path> <cert_path>",
+		ArgsUsage: "[<key_path> <cert_path>]",
 		Flags: []cli.Flag{
 			&cli.StringFlag{
 				Name:    "name",
 				Aliases: []string{"n"},
-				Usage:   "key name",
+				Usage:   "key name (required if --plugin is set)",
+			},
```
Comment (gokarnm) [inline cmd/notation/key.go:45]: The alpha 1 release supported signing only with a local key and cert file. For the newer `config.json` and CLI experience, key name (friendly name) should always be specified while adding a new key, as users will subsequently only refer to the key using its name for `notation sign` command.

File: cmd/notation/key.go [hunk 2]
```diff
@@ -71,63 +85,104 @@ var (
...
 	if err != nil {
-		return err
+		return nil, err
 	}
-	if err := cfg.Save(); err != nil {
-		return err
+	name := ctx.String("name")
+	if name == "" {
+		name = nameFromPath(keyPath)
 	}
```
Comment (gokarnm) [inline cmd/notation/key.go:165]: Ok, so we derive the name if not explicitly provided. IMO this make discovering the key name to be used difficult when user runs `notation sign --keyName somename` command. How does the user know what to set for `keyName` argument? We should probably remove deriving the name from the path.

File: cmd/notation/verify.go [hunk 3]
```diff
@@ -137,11 +138,11 @@ func appendCertPathFromName(paths, names []string) ([]string, error) {
...
-		path, ok := cfg.VerificationCertificates.Certificates.Get(name)
-		if !ok {
+		idx := slices.Index(cfg.VerificationCertificates.Certificates, name)
+		if idx < 0 {
 			return nil, errors.New("verification certificate not found: " + name)
 		}
-		paths = append(paths, path)
+		paths = append(paths, cfg.VerificationCertificates.Certificates[idx].Path)
 	}
 	return paths, nil
```
Comment (gokarnm) [inline cmd/notation/verify.go:147]: fyi, VerificationCertificates though config.json will go away when we implement updated [trust store](https://github.com/notaryproject/notaryproject/blob/main/trust-store-trust-policy-specification.md#trust-store) and [trust policy](https://github.com/notaryproject/notaryproject/blob/main/trust-store-trust-policy-specification.md#trust-store) spec.

File: go.sum [hunk 4]
```diff
@@ -99,6 +96,8 @@ github.com/prometheus/common v0.6.0/go.mod h1:eBmuwkDJBwy6iBfxCBob6t6dR6ENT/y+J+
 github.com/prometheus/procfs v0.0.0-20181005140218-185b4288413d/go.mod h1:c3At6R/oaqEKCNdg8wHV1ftS6bRYblBhIjjI8uT2IGk=
 github.com/prometheus/procfs v0.0.2/go.mod h1:TjEm7ze935MbeOT/UhFTIMYKhuLP4wbCsTZCD3I8kEA=
 github.com/prometheus/procfs v0.0.3/go.mod h1:4A/X28fw3Fc593LaREMrKMqOKvUAntwMDaekg4FpcdQ=
+github.com/qmuntal/notation-go v0.7.0-alpha.1.0.20220429152921-f0636e164767 h1:jgS3O1+FduH9ywKWadFSFcnNIcdbuJmqUV2RFMDCv30=
+github.com/qmuntal/notation-go v0.7.0-alpha.1.0.20220429152921-f0636e164767/go.mod h1:B/JbgikwvLoD9gFeWYhFtxIEozRoFAs9q31mj3vz1i4=
```
Comment (gokarnm) [inline go.sum:100]: Assuming this will be updated once plugin manager updates are merged to main.

**Review by gokarnm — APPROVED** [review]
LGTM. Have a couple of minor comments. I'd ask a review from @shizhMSFT too as he wrote most of this code.

File: cmd/notation/key.go (hunk 2 above)
Comment (qmuntal) [reply cmd/notation/key.go:165]: Agree, this was confusing when I first used the command without specifying the key name.

File: go.sum (hunk 4 above)
Comment (qmuntal) [reply go.sum]: I've changed notation-go reference so it points to `https://github.com/notaryproject/notation@main`, as the necessary functionality has already been merged.

File: cmd/notation/key.go (hunk 2 above)
Comment (gokarnm) [reply cmd/notation/key.go:165]: Do we want to make name a required CLI argument then, instead of inferring it, I'd prefer that.

Comment (qmuntal) [reply cmd/notation/key.go]: Done in [d1bb9be](https://github.com/notaryproject/notation/pull/168/commits/d1bb9be0b6b1fa883bcfe850e2b6568bf8b6ae0e)

File: cmd/notation/key.go [hunk 5]
```diff
@@ -71,63 +86,97 @@ var (
...
-func addKeyCore(cfg *config.File, name, keyPath, certPath string, markDefault bool) (bool, error) {
-	if ok := cfg.SigningKeys.Keys.Append(name, keyPath, certPath); !ok {
-		return false, errors.New(name + ": already exists")
+func addExternalKey(ctx *cli.Context, pluginName, keyName string) (*config.KeySuite, error) {
+	id := ctx.String("id")
+	if id == "" {
+		return nil, errors.New("missing key id")
+	}
+	mgr := manager.NewManager()
+	p, err := mgr.Get(context.Background(), pluginName)
```
Comment (shizhMSFT) [inline cmd/notation/key.go:130]:
```suggestion
	p, err := mgr.Get(ctx.Context, pluginName)
```

File: cmd/notation/key.go [hunk 6]
```diff
@@ -71,63 +86,97 @@ var (
...
-		fmt.Println(name)
+		fmt.Println(key.Name)
 	}
 	return nil
 }
 
-func addKeyCore(cfg *config.File, name, keyPath, certPath string, markDefault bool) (bool, error) {
-	if ok := cfg.SigningKeys.Keys.Append(name, keyPath, certPath); !ok {
-		return false, errors.New(name + ": already exists")
+func addExternalKey(ctx *cli.Context, pluginName, keyName string) (*config.KeySuite, error) {
```
Comment (shizhMSFT) [inline cmd/notation/key.go:124]: nit: Will `config.KeySuite` be better than `*config.KeySuite`?

File: cmd/notation/key.go [hunk 7]
```diff
@@ -168,7 +217,7 @@ func listKeys(ctx *cli.Context) error {
 	}
 
 	// write out
-	printKeySet(cfg.SigningKeys.Default, cfg.SigningKeys.Keys)
+	ioutil.PrintKeyMap(os.Stdout, cfg.SigningKeys.Default, cfg.SigningKeys.Keys)
 	return nil
```
Comment (shizhMSFT) [inline cmd/notation/key.go:221]:
```suggestion
	return ioutil.PrintKeyMap(os.Stdout, cfg.SigningKeys.Default, cfg.SigningKeys.Keys)
```

File: cmd/notation/cert.go [hunk 8]
```diff
@@ -137,7 +142,7 @@ func listCerts(ctx *cli.Context) error {
 	}
 
 	// write out
-	printCertificateSet(cfg.VerificationCertificates.Certificates)
+	ioutil.PrintCertificateMap(os.Stdout, cfg.VerificationCertificates.Certificates)
 	return nil
```
Comment (shizhMSFT) [inline cmd/notation/cert.go:146]:
```suggestion
	return ioutil.PrintCertificateMap(os.Stdout, cfg.VerificationCertificates.Certificates)
```

**Review by shizhMSFT — CHANGES_REQUESTED** [review]

File: cmd/notation/cert.go [hunk 9]
```diff
@@ -59,9 +62,10 @@ var (
 		ArgsUsage: "<host> ...",
 		Flags: []cli.Flag{
 			&cli.StringFlag{
-				Name:    "name",
-				Aliases: []string{"n"},
-				Usage:   "key and certificate name",
+				Name:     "name",
+				Aliases:  []string{"n"},
+				Usage:    "key and certificate name",
+				Required: true,
```
[the remark this answers is not in the document]
Comment (qmuntal) [reply cmd/notation/cert.go:68]: Yep, I set it in the wrong command.

File: cmd/notation/key.go (hunk 6 above)
Comment (qmuntal) [reply cmd/notation/key.go:124]: Yep.

File: cmd/notation/plugin.go [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+	pluginListCommand = &cli.Command{
+		Name:   "list",
+		Usage:  "List registered plugins",
+		Action: listPlugins,
+	}
+)
+
+func listPlugins(ctx *cli.Context) error {
+	mgr := manager.NewManager()
```
[the remark this answers is not in the document]
Comment (qmuntal) [reply cmd/notation/plugin.go:30]: I initially wanted to move the `config` package to `notation-go`, so `NewManager` could take the path from that package without requiring the user to set it. In fact, @gokarnm also made this same comment in a previous PR.

I'm no longer sure moving `config` to `notation-go` is a good idea, at least for the short term, so I'll modify add the `root` parameter to `NewManager` on the next `notation-go` PR, which is https://github.com/notaryproject/notation-go/pull/42.

File: internal/cmd/signer.go [hunk 11]
```diff
@@ -18,11 +19,13 @@ func GetSigner(ctx *cli.Context) (notation.Signer, error) {
...
 	} else {
-		var err error
-		keyPath, certPath, err = config.ResolveKeyPath(ctx.String(FlagKey.Name))
+		key, err := config.ResolveKey(ctx.String(FlagKey.Name))
 		if err != nil {
 			return nil, err
 		}
+		if key.X509KeyPair == nil {
+			return nil, errors.New("invalid key type")
+		}
```
[the remark this answers is not in the document]
Comment (qmuntal) [reply internal/cmd/signer.go:28]: Plugin signers are still not in `notation-go@main`. I'll update this code once https://github.com/notaryproject/notation-go/pull/42 is merged.

File: internal/slices/slices.go [hunk 12]
```diff
@@ -0,0 +1,27 @@
+package slices
```
[the remark this answers is not in the document]
Comment (qmuntal) [reply internal/slices/slices.go:1]: Done!

File: pkg/config/config.go [hunk 13]
```diff
@@ -6,6 +6,38 @@ import (
 	"path/filepath"
 )
 
+type X509KeyPair struct {
```
[the remark this answers is not in the document]
Comment (qmuntal) [reply pkg/config/config.go:9]: Done!

File: pkg/config/config.go [hunk 14]
```diff
@@ -6,6 +6,38 @@ import (
...
+	KeyPath         string `json:"keyPath"`
+	CertificatePath string `json:"certPath"`
+}
+
+type ExternalKey struct {
+	ID         string `json:"id"`
+	PluginName string `json:"pluginName"`
+}
```
[the remark this answers is not in the document]
Comment (qmuntal) [reply pkg/config/config.go:17]: Done!

File: pkg/config/util.go [hunk 15]
```diff
@@ -27,33 +26,19 @@ func IsRegistryInsecure(target string) bool {
...
-	}
-	path, ok := config.VerificationCertificates.Certificates.Get(name)
-	if !ok {
-		return "", ErrCertificateNotFound
+	idx := slices.Index(config.SigningKeys.Keys, name)
+	if idx < 0 {
+		return KeySuite{}, ErrKeyNotFound
 	}
-	return path, nil
+	return config.SigningKeys.Keys[idx], ErrKeyNotFound
```
Comment (shizhMSFT) [inline pkg/config/util.go:43]:
```suggestion
	return config.SigningKeys.Keys[idx], nil
```

**Review by shizhMSFT — COMMENTED** [review]
I've also reopened some comments in the previous review as they are marked resolved without changes.

**Comment by qmuntal** [comment]
> I've also reopened some comments in the previous review as they are marked resolved without changes.

@shizhMSFT I think I got them all now.

**Review by shizhMSFT — APPROVED** [review]
LGTM with one unresolved comment.

**Comment by qmuntal** [comment]
> LGTM with one unresolved comment.

Resolved

**Pull request merged**